CIP-ledger-evolution/STS.md

```diff
@@ -0,0 +1,195 @@
+# The Small Step Semantics Framework for Cardano
+
+This document is heavily based on
+an [original version](https://github.com/input-output-hk/cardano-ledger/blob/master/docs/small-step-semantics/small-step-semantics.tex)
+by [Nicholas Clarke](https://github.com/nc6).
+
+## Introduction
+
+In defining the Cardano ledger, we are concerned with the means to construct inductive datatypes
+satisfying some validity conditions. For example, we wish to consider when a sequence of transactions
+forms a valid ledger, or a sequence of blocks forms a valid chain.
+This document describes the methods we use to describe such validity conditions and how
+they result in the construction of valid states.
+
+## Transition Systems
+
+In describing a state transition system `L` we are principally concerned with six things:
+
+* **States** - The underlying datatype of our system.
+
+* **Transitions** - The means by which we might move from one (valid) state to another.
+
+* **Signals** - The means by which transitions are triggered.
+
+* **Predicates** - We do not prescribe a formal language with which to define predicates,
+but rather leave this up to the description of specific systems.
+Even so, a formal language may not been needed as long as the notation is clear and precise enough.
+Typically, finite set theory is all that is needed, with notation for concepts such as
+integers, sequences, powersets, functions, maps, comprehensions, conditionals, etc.
+Most of the predicates will have a value of true or false,
+but we will stretch the rules later when we define the composition of transition systems.
+
+* **Rules** - Judgments describing valid transitions. A rule has an antecedent
+(sometimes called premise) and a consequent (sometimes called conclusion), such that if
+the predicates in the antecedent hold, the consequent is assumed to be a valid state or transition.
+The predicates in the antecedent may contain variables for the state, the signal,
+and the environment (see below).
+
+* **Environment** - Sometimes we may implicitly rely on some additional information being present
+in the system to evaluate the rules. An environment does not have to exist (or, equivalently,
+we may have an empty environment), but crucially an environment is not modified by
+transitions.
+
+**Definition (State transition system)**
+
+A state transition system `L` is given by a 5-tuple `(S, T, Σ, R, Γ)`
+where:
+* `S` is a set of (not necessarily valid) states.
+* `Σ` is a set of signals.
+* `Γ` is a set of environment values.
+* `T` is a set of (not necessarily valid) transitions. We have that `T ⊆ 𝒫 (Γ × S × Σ × S)`.
```

Comment:
Suggestion: remove (`not necessarily valid`) as it's even more confusing in this case. The definition as a subset of cartesian product of other sets is enough to make it clear what is it

Reply:
I think the worry was that folks would equate the subset `T` with the set of valid transitions, as that was the reason for it being a subset. no?

Reply:
Well, the fact the transition relation is guarded by rules makes it clear that just `T` is not enough

Reply:
you define the set of `valid transitions` below, as the subset of `T` s.t there's a predicate `r` which holds.